subgraph/src/subscriptions.ts

```diff
@@ -44,12 +57,58 @@ export function handleSubscribe(event: SubscribeEvent): void {
 
   let user = loadOrCreateUser(event.params.user);
 
-  let sub = new ActiveSubscription(event.params.user);
-  sub.user = user.id;
-  sub.start = event.params.start;
-  sub.end = event.params.end;
-  sub.rate = event.params.rate;
-  sub.save();
+  let sub = ActiveSubscription.load(event.params.user);
+  if (sub == null) {
+    sub = new ActiveSubscription(event.params.user);
+    sub.user = user.id;
+    sub.start = event.params.start;
+    sub.end = event.params.end;
+    sub.rate = event.params.rate;
+    sub.save();
+    // Since ActiveSubscription record does not exist, the user is subscribing for the 1st time.
+    // Create and store a UserSubscriptionCreatedEvent record.
+    buildAndSaveUserSubscriptionCreatedEvent(user, sub, event);
+    // If the user calls the subscribe function on the contract and the ActiveSubscription.end > block.timestamp,
+    // then the contract will call the unsubscribe function, which emits the Unsubscribe event.
+    // In the `handleUnsubscribe` fn below, we create a `UserSubscriptionCanceledEvent` record on the Unsubscribe event.
+    // The contract then recreates the ActiveSubscription and emits a Subscribe event; which is handled by this function.
+    // In this instance where an Unsubscribe is immediately followed by a Subscribe,
+    // the user did not intend to "Cancel" their subscription, they meant to renew it.
+    // As a result, find the created `UserSubscriptionCanceledEvent` record for the user and remove it from the store,
+    // and create a `UserSubscriptionRenewalEvent` record.
+    // NOTE:  preferably, we don't create the `UserSubscriptionCanceledEvent` record, but not sure how to ensure that.
+    //        alternatively, we create the `UserSubscriptionCanceledEvent` record but don't remove it to track that it occurred.
+    let canceledEventId = buildUserSubscriptionEventId(
+      user.id,
+      USER_SUBSCRIPTION_EVENT_TYPE__CANCELED,
+      event.block.timestamp
+    );
+    let canceledEvent = UserSubscriptionCanceledEvent.load(canceledEventId);
+    if (canceledEvent != null) {
+      buildAndSaveUserSubscriptionRenewalEvent(user, sub, event);
+
+      store.remove(
+        'UserSubscriptionCanceledEvent',
+        canceledEventId.toHexString()
+      );
+    }
```

Comment:
This is an open question for this PR. Do we think it is better to:
1. Create the `UserSubscriptionCanceledEvent` and then remove it if the user was trying to renew their subscription and not cancel it
2. Keep it as a representation of the events that occurred
3. Another option...

Thoughts?

Reply:
I think we'd rather clear the `Canceled`. Is the order of calls to `handleSubscribe` and `handleUnsubscribe` always the same? I'm worried we might need to consider a case of `[Subscribe, Unsubscribe]` and `[Unsubscribe, Subscribe]` where just a lone `[Unsubscribe]` becomes the `Canceled`.

Reply:
I agree with clearing the `Canceled` event if the user is renewing. So I will keep that logic.

> Is the order of calls to handleSubscribe and handleUnsubscribe always the same?

The contract always emits them in this order. But I do not know/think that we can guarantee that the mapping function will always be invoked in that order. I will also check for a `Created` event in the same timestamp in the unsubscribe code before creating the `Canceled` event there, just in case. I think that is the best due-diligence we can do